### DISCUSSION TITLE
Can i use PnP core sdk inside .net 8.0 console application and use AD App registration which uses self-signed certificate

### DISCUSSION THREAD

**Discussion by mvcsharepointdev** [pnp/pnpcore#1382] [Q&A]
I have create a AD app registration which uses self-signed certificate for authentication. and i grant the AD app registration full control on SharePoint and graph APIs.

now i want to build a .net 8.0 console application which run on schedule bases .. so can i use PnP core SDK with AD app that uses self-signed certificate ? if this is possible, then is there a sample code that can help me?

Thanks

**Comment by wizofaus** [comment]
Assuming you have an X509Certificate2 object loaded from a file or embedded resource etc., you can just pass a X509CertificateAuthenticationProvider to PnpContextFactory.CreateAsync().

**Reply by mvcsharepointdev** [reply]
@wizofaus any sample code please, thanks

**Reply by mvcsharepointdev** [reply]
@wizofaus here what i tried so far, but got error:-

```
using PnP.Core.Auth;
using PnP.Core.Services;
using System.Security.Cryptography.X509Certificates;

namespace ConsoleApp3
{
    public class Program
    {
        private readonly IPnPContextFactory pnpContextFactory;
        public Program(IPnPContextFactory pnpContextFactory)
        {
            this.pnpContextFactory = pnpContextFactory;

        }
        static async Task Main(string[] args)
        {
            var tenantId = "your-tenant-id";
            var clientId = "your-application-id";
            var certificatePath = "path-to-your-certificate.pfx";
            var certificatePassword = "your-certificate-password";

            // Load the certificate
            var authenticationProvider = new X509CertificateAuthenticationProvider(tenantId, clientId, X509Certificate2.CreateFromPemFile(certificatePath, certificatePassword));

            // Use the authentication provider to configure PnPContextFactory
            var pnpContextFactory = new PnPContextFactory(new AuthenticationProviderFactory((resource) => authenticationProvider));

            // Create a PnPContext for your SharePoint site
            var siteUrl = "https://yourtenant.sharepoint.com/sites/yoursite";
            using var context = await pnpContextFactory.CreateAsync(new Uri(siteUrl));
        }
    }
}
```

![image](https://github.com/pnp/pnpcore/assets/79726511/7abe8cc5-7348-4bf6-a8f4-e70c5383352d)

any advice? Thanks

**Comment by wizofaus** [comment]
Just pass the X509CertificateAuthenticationProvider directly to CreateAsync( ), 2nd argument.  I don't think you need an AuthenticationProviderFactory.  I'm using a .NET Core API setup with DI etc. so it's slightly different, but I can't see why that wouldn't work.

**Reply by mvcsharepointdev** [reply]
@wizofaus thanks for the reply. it is weird that there is not any sample code on the net for such a requirement, i still can not see how i need to pass pass the X509CertificateAuthenticationProvider directly to CreateAsync( )? and how i can do so without the AuthenticationProviderFactory?

**Reply by Magnus-K-work** [reply]
Late reply, but maybe this can help someone.

The IPnPContextFactory.CreateAsync method can take two parameters where the second is an authentication provider, so just call it like this:
```
// Instantiate a certificate
var certificate = new X509Certificate2(<path to pfx certificate file>, <password for the certificate file>);

// Configure the interactive authentication provider as default
var authenticationProvider = new X509CertificateAuthenticationProvider(
    clientId: <client id>,
    tenantId: <tenant id>,
    certificate: certificate);

// Create a PnPContext
using (var context = await _pnpContextFactory.CreateAsync(new Uri(RootSiteUrl), authenticationProvider))
{
   // Do stuff
}
```

**Comment by wizofaus** [comment]
It's literally the 2nd parameter:
```  public Task<PnPContext> CreateAsync(Uri url, IAuthenticationProvider authenticationProvider, PnPContextOptions options = null);```

**Reply by mvcsharepointdev** [reply]
@wizofaus i am really getting confised, can you post a completed code, as i am not sure how to fit this code:-

`  public Task<PnPContext> CreateAsync(Uri url, IAuthenticationProvider authenticationProvider, PnPContextOptions options = null);`
inside my original code? Thanks

**Comment by wizofaus** [comment]
You're not, that's just the definition of CreateAsync. You just need to pass your authentication provider as the second parameter after the url.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jansenbe** [comment]
@mvcsharepointdev : see https://github.com/pnp/pnpcore/blob/dev/docs/polyglot/Getting%20started%20-%20application%20permissions.ipynb
